ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -33,119 +34,170 @@ import           Language.LSP.Server            (ProgressAmount (..),
+      (Nothing, _) -> modifyTVar' todoVar (+ 1)
+      (Just 0, 0)  -> pure ()
+      (Just 0, _)  -> modifyTVar' doneVar pred
+      (Just _, 0)  -> modifyTVar' doneVar (+ 1)
+      (Just _, _)  -> pure ()
   where
     alterPrevAndNew = do
-        prev <- Focus.lookup
-        Focus.alter alter
-        new <- Focus.lookupWithDefault 0
-        return (prev, new)
+      prev <- Focus.lookup
+      Focus.alter alter
+      new <- Focus.lookupWithDefault 0
+      return (prev, new)
     alter x = let x' = maybe (shift 0) shift x in Just x'
 
-progressReporting
-  :: Maybe (LSP.LanguageContextEnv c)
-  -> ProgressReportingStyle
-  -> IO ProgressReporting
-progressReporting Nothing _optProgressStyle = noProgressReporting
-progressReporting (Just lspEnv) optProgressStyle = do
-    inProgressState <- newInProgress
-    progressState <- newVar NotStarted
-    let progressUpdate event = updateStateVar $ Event event
-        progressStop  = updateStateVar StopProgress
-        updateStateVar = modifyVar_ progressState . updateState (lspShakeProgressNew inProgressState)
-        inProgress = updateStateForFile inProgressState
-    return ProgressReporting{..}
-    where
-        lspShakeProgressNew :: InProgressState -> IO ()
-        lspShakeProgressNew InProgressState{..} =
-            LSP.runLspT lspEnv $ withProgress "Processing" Nothing NotCancellable $ \update -> loop update 0
-            where
-                loop _ _ | optProgressStyle == NoProgress = forever $ liftIO $ threadDelay maxBound
-                loop update prevPct = do
-                    (todo, done, nextPct) <- liftIO $ atomically $ do
-                        todo <- readTVar todoVar
-                        done <- readTVar doneVar
-                        let nextFrac :: Double
-                            nextFrac = if todo == 0 then 0 else fromIntegral done / fromIntegral todo
-                            nextPct :: UInt
-                            nextPct = floor $ 100 * nextFrac
-                        when (nextPct == prevPct) retry
-                        pure (todo, done, nextPct)
-
-                    _ <- update (ProgressAmount (Just nextPct) (Just $ T.pack $ show done <> "/" <> show todo))
-                    loop update nextPct
-        updateStateForFile inProgress file = actionBracket (f succ) (const $ f pred) . const
-            -- This functions are deliberately eta-expanded to avoid space leaks.
-            -- Do not remove the eta-expansion without profiling a session with at
-            -- least 1000 modifications.
-            where
-              f shift = recordProgress inProgress file shift
-
-mRunLspT :: Applicative m => Maybe (LSP.LanguageContextEnv c ) -> LSP.LspT c m () -> m ()
+progressReporting ::
+  (MonadUnliftIO m, MonadIO m) =>
+  Maybe (LSP.LanguageContextEnv c) ->
+  T.Text ->
+  ProgressReportingStyle ->
+  IO (ProgressReporting m)
+progressReporting = progressReporting' newInProgress
+
+progressReportingOutsideState ::
```

Comment:
At the moment it's pretty unclear when you would use one of these or the other. Can we write down why in the haddock? I think we maybe also learned some other things that we should write down, like the stuff about whether `Stopped` means anything.

Reply:
done